refactor-spotify-api.md

```diff
@@ -0,0 +1,24 @@
+
+# Refactor Spofity API Layer + Create better user experience
+
+
+- Today when user enters the app he is prompted to click the "Start Jaming" button twice.
+- This user experience, is a cause of an implementation issue. In other words, not ideal code design.
+- And after the first redirect we are prompted with the same message again, altho we laready dont need to click twice.
+- Technicaly we need 1. redirect the user to spotify url. 2. grab some token from their redirect, for future use.
+- Right now it all happens in the same function `getUserAccessToken` which is called on "Start Jaming" click.
+- When thinking about the optimal user experice, there should be only one "Start Jaming" click.
+- After which this button should go away, and we should see a new screen. 
+- There should be a conditioning logic in `App` that decides whether it renders the "Start Jaming" button, or the rest of the UI.
+- This conditioning logic should be based on the state of a dedicated boolean somthing like `hasUserLoggedIn`
+- In Spofity.js there is a function called `getUserAccessToken` which should be renamed and refactored, or even removed.
+- Naming; `getUserAccessToken` suggest that the function gets something, altho it implicitly "gets the redirect url" with the token, in reality no one calling it uses what it gets as a return value.
+- Logic; main business logic is to redirect. In this case, its a very simple fn.
+- You can even export the redirect url from Spotify.js and import it in App.js and pass it to the href. thats it.
+- Now we need to add some state to the App component which will hold `hasUserLoggedIn` and the `token`
```

Comment:
I'm trying to make sure I got it right, after reading your code review and thinking of it a little bit:

I was thinkig, instead of adding the `hasUserLoggedIn` to `App.state`, pushing the `userAccessToken` itself to `App.state`.
Starting with an empty string, which is falsey:
`if (!App.state.userAccessToken) {
render the Start Jammming button and that's it
} else{ 
render the entier App
}`
It should save some coding time but I'm not sure it will work better.
Would be happy to hear your insights.

Reply:
what you suggest make sense, less one variable to declare. 
The only downside, which is not huge at all in this case, is naming, `hasUserLoggedIn` is more explicit/descriptive to whats going on in the user state. While `userAccessToken` is more implicit to why we render the button. you can always rename for the context, meaning

```

render() {
  const { userAccessToken: hasUserLoggedIn } = this.state

  return (
    <div>
      {
        hasUserLoggedIn
          ? (
             <Jamming/>
          )
          : (
             <Enter/>
           )
      }
    </div>
  )
}

```